q2_metadata/normalization/_norm_errors.py

```diff
@@ -0,0 +1,204 @@
+    """Exception raised for error due to the
+    presence of a rule that's not recognized."""
+
+    def __init__(self, rule, variable, value):
+        self.rule = rule
+        self.variable = variable
+        self.value = value
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> rule not recognized' % message
+
+
+class ExpectedError(Error):
+    """Exception raised for errors in the
+    formatting of the "expected values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'expected'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
+        if not isinstance(self.value, list):
+            return 'is not a list'
+        str_items = [x for x in self.value if isinstance(x, str)]
+        if len(str_items) != len(self.value):
+            return 'not all items are strings'
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> %s' % (message, self.message)
+
+
+class OntologyError(Error):
+    """Exception raised for errors in the
+    formatting of the "ontology values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'ontology'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
+        if not isinstance(self.value, str):
+            return 'is not a string'
+        ontologies = ['Gazetteer ontology']
+        if self.value not in ontologies:
+            return 'none of %s' % ', '.join(ontologies)
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> %s' % (message, self.message)
+
+
+class RemapError(Error):
+    """Exception raised for errors in the
+    formatting of the "remap values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'remap'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
+        if not isinstance(self.value, dict):
+            return 'is not a dictionary'
+        str_values = [x for x, y in self.value.items() if isinstance(y, str)]
+        if len(str_values) != len(self.value.keys()):
+            return 'not all dictionary values are strings'
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> %s' % (message, self.message)
+
+
+class ValidationError(Error):
+    """Exception raised for errors in the
+    formatting of the "validation values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'validation'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
```

Comment:
what is this doing?

Reply:
getting a message for the user to know why his formatting is wrong.